certbot-nginx/certbot_nginx/_internal/configurator.py

```diff
@@ -1189,10 +1189,13 @@ def perform(self, achalls: List[achallenges.AnnotatedChallenge]
 
         """
         self._chall_out += len(achalls)
-        responses: List[Optional[challenges.HTTP01Response]] = [None] * len(achalls)
+        responses: List[Optional[challenges.ChallengeResponse]] = [None] * len(achalls)
         http_doer = http_01.NginxHttp01(self)
 
-        for i, achall in enumerate(achalls):
+        key_achalls = [achall for achall in achalls
+                       if isinstance(achall, achallenges.KeyAuthorizationAnnotatedChallenge)]
+
+        for i, achall in enumerate(key_achalls):
             # Currently also have chall_doer hold associated index of the
             # challenge. This helps to put all of the responses back together
             # when they are all complete.
```

Comment:
nit: can we change this to something like:
```suggestion
        for i, achall in enumerate(achalls):
            # Currently also have chall_doer hold associated index of the
            # challenge. This helps to put all of the responses back together
            # when they are all complete.
            assert isinstance(achall, achallenges.KeyAuthorizationAnnotatedChallenge)
```
We could also raise some type an error in the body of an `if` statement if you prefer. This error should never happen, I'd just like to avoid us silently ignoring unexpected challenge types or iterating over `achalls` twice.

Reply:
I did that, with an `if` branch and an appropriate error raised in this case.